particle-iot/device-os
# [system] close cloud socket on errors

### Problem

Cloud socket is not closed on communication errors. With heavy traffic unread data may be left in it, which may prevent Gen 3 devices from establishing PPP link or performing network communications due to lack of TX/RX buffers.

### Solution

Close the cloud socket on errors.

Caveat: SARA R4 internal TCP/IP stack (used on Electrons) leaks sockets (opening/closing them too often will eventually prevent us from creating any more sockets). We leave the existing behavior on Electrons.

### Steps to Test

Run the following app on cellular Gen 3 device. Eventually after successfully publishing for a while, the device may get into a blinking green state. It should recover successfully.

```c++
#include "application.h"
SerialLogHandler logHandler(LOG_LEVEL_ALL);
void setup() {
}

const int num_runs = 10;

void loop() {
    static system_tick_t last = 0;
    for (int i=0; i < num_runs; i++) {
        delay(250);
        Particle.publish("dummy", "123");
        if (millis() - last >= 1000) {
            last = millis();
            int32_t value = -1;
            AbstractIntegerDiagnosticData::get(DIAG_ID_CLOUD_RATE_LIMITED_EVENTS, value);
            Log.trace("rate limited: %d", value);
        }
    }
}
```

### References

- [CH50189]

---

### Completeness
